// facturacionelectronicapy-xmlgen/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const jsonDeMain_service_1 = __importDefault(require("./services/jsonDeMain.service"));
const jsonEventoMain_service_1 = __importDefault(require("./services/jsonEventoMain.service"));
class DE {
    constructor() {
        this.generateXMLDE = (params, data, config) => {
            return jsonDeMain_service_1.default.generateXMLDE(params, data, config);
        };
        this.generateXMLEventoCancelacion = (id, params, data, config) => {
            return jsonEventoMain_service_1.default.generateXMLEventoCancelacion(id, params, data, config);
        };
        this.generateXMLEventoInutilizacion = (id, params, data, config) => {
            return jsonEventoMain_service_1.default.generateXMLEventoInutilizacion(id, params, data, config);
        };
        this.generateXMLEventoConformidad = (id, params, data, config) => {
            return jsonEventoMain_service_1.default.generateXMLEventoConformidad(id, params, data, config);
        };
        this.generateXMLEventoDisconformidad = (id, params, data, config) => {
            return jsonEventoMain_service_1.default.generateXMLEventoDisconformidad(id, params, data, config);
        };
        this.generateXMLEventoDesconocimiento = (id, params, data, config) => {
            return jsonEventoMain_service_1.default.generateXMLEventoDesconocimiento(id, params, data, config);
        };
        this.generateXMLEventoNotificacion = (id, params, data, config) => {
            return jsonEventoMain_service_1.default.generateXMLEventoNotificacion(id, params, data, config);
        };
        this.generateXMLEventoNominacion = (id, params, data, config) => {
            return jsonEventoMain_service_1.default.generateXMLEventoNominacion(id, params, data, config);
        };
        this.generateXMLEventoActualizacionDatosTransporte = (id, params, data, config) => {
            return jsonEventoMain_service_1.default.generateXMLEventoActualizacionDatosTransporte(id, params, data, config);
        };
        this.consultarPaises = () => {
            return new Promise((resolve, reject) => {
                //Enviar Copia
                let paises = [];
                for (let index = 0; index < jsonDeMain_service_1.default.getPaises().length; index++) {
                    const pais = jsonDeMain_service_1.default.getPaises()[index];
                    paises.push(Object.assign({}, pais));
                }
                resolve(paises);
            });
        };
        this.consultarDepartamentos = () => {
            return new Promise((resolve, reject) => {
                //Enviar Copia
                let departamentos = [];
                for (let index = 0; index < jsonDeMain_service_1.default.getDepartamentos().length; index++) {
                    const dep = jsonDeMain_service_1.default.getDepartamentos()[index];
                    departamentos.push(Object.assign({}, dep));
                }
                resolve(departamentos);
            });
        };
        this.consultarDistritos = (departamento) => {
            return new Promise((resolve, reject) => {
                //Enviar Copia
                let distritos = [];
                for (let index = 0; index < jsonDeMain_service_1.default.getDistritos(departamento).length; index++) {
                    const dis = jsonDeMain_service_1.default.getDistritos(departamento)[index];
                    distritos.push(Object.assign({}, dis));
                }
                resolve(distritos);
            });
        };
        this.consultarCiudades = (distrito) => {
            return new Promise((resolve, reject) => {
                //Enviar Copia
                let ciudades = [];
                for (let index = 0; index < jsonDeMain_service_1.default.getCiudades(distrito).length; index++) {
                    const ciu = jsonDeMain_service_1.default.getCiudades(distrito)[index];
                    ciudades.push(Object.assign({}, ciu));
                }
                resolve(ciudades);
            });
        };
        this.consultarTiposRegimenes = () => {
            return new Promise((resolve, reject) => {
                let tiposRegimenes = [];
                for (let index = 0; index < jsonDeMain_service_1.default.getTiposRegimenes().length; index++) {
                    const tip = jsonDeMain_service_1.default.getTiposRegimenes()[index];
                    tiposRegimenes.push(Object.assign({}, tip));
                }
                resolve(tiposRegimenes);
            });
        };
        this.getDepartamento = (departamentoId) => {
            let departamentos = jsonDeMain_service_1.default.getDepartamento(departamentoId);
            if (departamentos.length > 0) {
                return Object.assign({}, departamentos[0]);
            }
            else {
                return null;
            }
        };
        this.getDistrito = (distritoId) => {
            let distritos = jsonDeMain_service_1.default.getDistrito(distritoId);
            if (distritos.length > 0) {
                return Object.assign({}, distritos[0]);
            }
            else {
                return null;
            }
        };
        this.getCiudad = (ciudadId) => {
            let ciudades = jsonDeMain_service_1.default.getCiudad(ciudadId);
            if (ciudades.length > 0) {
                return Object.assign({}, ciudades[0]);
            }
            else {
                return null;
            }
        };
    }
}
exports.default = new DE();
