// facturacionelectronicapy-setapi/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const SET_1 = __importDefault(require("./SET"));
class SetAPI {
    constructor() {
        this.consulta = (id, cdc, env, cert, key, config) => {
            return SET_1.default.consulta(id, cdc, env, cert, key, config);
        };
        this.consultaRUC = (id, ruc, env, cert, key, config) => {
            return SET_1.default.consultaRUC(id, ruc, env, cert, key, config);
        };
        this.consultaLote = (id, numeroLote, env, cert, key, config) => {
            return SET_1.default.consultaLote(id, numeroLote, env, cert, key, config);
        };
        this.recibe = (id, xml, env, cert, key, config) => {
            return SET_1.default.recibe(id, xml, env, cert, key, config);
        };
        this.recibeLote = (id, xml, env, cert, key, config) => {
            return SET_1.default.recibeLote(id, xml, env, cert, key, config);
        };
        this.evento = (id, xml, env, cert, key, config) => {
            return SET_1.default.evento(id, xml, env, cert, key, config);
        };
    }
}
exports.default = new SetAPI();
