// facturacionelectronicapy-xmlsign/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const XMLDsigJava_1 = __importDefault(require("./XMLDsigJava"));
const XMLDsigNode_1 = __importDefault(require("./XMLDsigNode"));
class DESign {
    constructor() {
        this.signXML = (xml, file, password, signByNodeJS) => {
            if (signByNodeJS == true) {
                return XMLDsigNode_1.default.signDocument(xml, "DE", file, password);
            }
            else {
                return XMLDsigJava_1.default.signDocument(xml, "DE", file, password);
            }
            //return dsigJava.signDocument(xml, "DE", file, password);
        };
        this.signXMLFiles = (xmls, file, password, signByNodeJS) => {
            if (signByNodeJS == true) {
                return XMLDsigNode_1.default.signDocuments(xmls, "DE", file, password);
            }
            else {
                return XMLDsigJava_1.default.signDocuments(xmls, "DE", file, password);
            }
        };
        /*signXMLFilesNode = (
          xmls: Array<any>,
          file: any,
          password: any
        ): Promise<any> => {
          //return dsigJava.signDocuments(xmls, "DE", file, password);
          return dsigNode.signDocument(xmls, "DE", file, password);
        };*/
        this.signXMLEvento = (xml, file, password, signByNodeJS) => {
            if (signByNodeJS == true) {
                return XMLDsigNode_1.default.signEvento(xml, "rEve", file, password);
            }
            else {
                return XMLDsigJava_1.default.signEvento(xml, "rEve", file, password);
            }
            //return dsigNode.signDocument([xml], "DE", file, password);
        };
        this.signXMLRecibo = (xml, file, password, signByNodeJS) => {
            if (signByNodeJS == true) {
                return XMLDsigNode_1.default.signDocument(xml, "recibo", file, password);
            }
            else {
                return XMLDsigJava_1.default.signDocument(xml, "recibo", file, password);
            }
            //return dsigNode.signDocument([xml], "DE", file, password);
        };
        this.getExpiration = (file, password, useNodeJS) => {
            if (useNodeJS == true) {
                return XMLDsigNode_1.default.getExpiration(file, password);
            }
            else {
                return XMLDsigJava_1.default.getExpiration(file, password);
            }
        };
    }
}
exports.default = new DESign();
