// facturacionelectronicapy-xmlsign/dist/PKCS12.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const node_forge_1 = __importDefault(require("node-forge"));
class PKCS12 {
    openCertificate(file) {
        if (fs_1.default.existsSync(file)) {
            const pkcs12 = fs_1.default.readFileSync(file);
            this.p12Asn1 = node_forge_1.default.asn1.fromDer(pkcs12.toString("binary"));
        }
        else {
            throw Error(file + " no encontrado!");
        }
    }
    openFile(file, passphase) {
        this.openCertificate(file);
        this.p12 = node_forge_1.default.pkcs12.pkcs12FromAsn1(this.p12Asn1, false, passphase);
    }
    clean() {
        this.p12 = undefined;
    }
    getPrivateKey() {
        for (let i = 0; i < this.p12.safeContents.length; i++) {
            if (this.p12.safeContents[i].safeBags[0].key) {
                return node_forge_1.default.pki.privateKeyToPem(this.p12.safeContents[i].safeBags[0].key);
            }
        }
        return null;
    }
    getCertificate() {
        for (let i = 0; i < this.p12.safeContents.length; i++) {
            if (this.p12.safeContents[i].safeBags[0].cert) {
                const b64 = node_forge_1.default.pki.certificateToPem(this.p12.safeContents[i].safeBags[0].cert);
                const l = b64.split("\n");
                l.pop();
                l.pop();
                l[0] = "";
                return l.join("\n");
            }
        }
        return null;
    }
    signature(xml, privateKey) {
        //const md = forge.md.sha256.create();
        const md = node_forge_1.default.md.sha256.create();
        md.update(xml, "utf8");
        var key = node_forge_1.default.pki.privateKeyFromPem(privateKey);
        return key.sign(md);
    }
    getPrivateKeyFromPEM(privateKey) {
        var key = node_forge_1.default.pki.privateKeyFromPem(privateKey);
        return key;
    }
    signature22222(xml, privateKey) {
        const md = node_forge_1.default.md.sha256.create();
        md.update(xml, "utf8");
        var key = node_forge_1.default.pki.privateKeyFromPem(privateKey);
        //let signature = forge.util.encode64(key.sign(md));
        return key.sign(md);
    }
}
exports.default = new PKCS12();
